File: 백준/Gold/1202. 보석 도둑/보석 도둑.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;

public class Main {
    private static class Jewel implements Comparable<Jewel> {
        int weight, cost;
        Jewel(int weight, int cost) {
            this.weight = weight;
            this.cost = cost;
        }
        @Override
        public int compareTo(Jewel other) {
            return this.weight - other.weight;
        }
        @Override
        public String toString() {
            return "weight : " + weight + " cost : " + cost;
        }
    }
    private static List<Integer> bags = new ArrayList<>();
    private static PriorityQueue<Jewel> jewels = new PriorityQueue<>();
    private static PriorityQueue<Integer> costs = new PriorityQueue<>(Collections.reverseOrder());
    private static long total = 0;
    public static void main(String[] args) throws IOException {
        int N = readInt(), K = readInt();
        for (int i = 0; i < N; i++) {
            jewels.add(new Jewel(readInt(), readInt()));
        }
        for (int i = 0; i < K; i++) {
            bags.add(readInt());
        }
        Collections.sort(bags);
        for (int weight : bags) {
            while (!jewels.isEmpty() && jewels.peek().weight <= weight) {
                costs.add(jewels.poll().cost);
            }
            if (!costs.isEmpty()) {
                total += costs.poll();
            }
        }
        System.out.println(total);
    }
    private static int readInt() throws IOException {
        int val = 0;
        int c = System.in.read();
        while (c <= ' ') {
            c = System.in.read();
        }
        do {
            val = 10 * val + c - 48;
        } while ((c = System.in.read()) >= 48 && c <= 57);
        return val;
    }
}

File: 백준/Gold/1202. 보석 도둑/보석 도둑.cc
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>

#define FAST_IO cin.tie(NULL)->ios::sync_with_stdio(false);

using namespace std;

struct Jewel {
    int weight, cost;
};

int main() {
    FAST_IO

    int N, K; cin >> N >> K;
    vector<Jewel> jewels(N);
    for (int i = 0; i < N; i++) {
        cin >> jewels[i].weight >> jewels[i].cost;
    }

    vector<int> bags(K);
    for (int i = 0; i < K; i++) {
        cin >> bags[i];
    }

    sort(jewels.begin(), jewels.end(), [](const Jewel &a, const Jewel &b) {
        return a.weight < b.weight;
    });
    sort(bags.begin(), bags.end());

    priority_queue<int> costs;
    long long total = 0;
    int idx = 0;
    for (int weight : bags) {
        while (idx < N && jewels[idx].weight <= weight) {
            costs.push(jewels[idx].cost);
            idx++;
        }
        if (!costs.empty()) {
            total += costs.top();
            costs.pop();
        }
    }

    cout << total << "\n";

    return EXIT_SUCCESS;
}